src/ui/components/SecondaryToolbox/ReduxDevTools.tsx

```diff
@@ -70,6 +79,115 @@ export const ReduxDevToolsPanel = () => {
+}
+
+const reduxDispatchJumpLocationCache = createCache<
+  [
+    replayClient: ReplayClientInterface,
+    point: ExecutionPoint,
+    time: number,
+    sourcesState: SourcesState
+  ],
+  PointDescription | undefined
+>({
+  config: { immutable: true },
+  debugLabel: "ReduxDispatchJumpLocation",
+  getKey: ([replayClient, point, time, sourcesState]) => point,
+  load: async ([replayClient, point, time, sourcesState]) => {
+    const pauseId = await pauseIdCache.readAsync(replayClient, point, time);
+    const frames = (await getPauseFramesAsync(replayClient, pauseId, sourcesState)) ?? [];
+    const filteredPauseFrames = frames.filter(frame => {
+      const { source } = frame;
+      if (!source) {
+        return false;
+      }
+
+      // Filter out everything in `node_modules`, so we have just app code left
+      // TODO There may be times when we care about renders queued by lib code
+      // TODO See about just filtering out React instead?
+      return !MORE_IGNORABLE_PARTIAL_URLS.some(partialUrl => source.url?.includes(partialUrl));
+    });
+
+    // The first 2 elements in filtered pause frames are from replay's redux stub, so they should be ignored
+    // The 3rd element is the user function that calls it, and will most likely be the `dispatch` call
+    let currentPreferredIndex = 2;
+    let preferredFrame = filteredPauseFrames[currentPreferredIndex];
+    let isMiddleware = true;
+
+    while (isMiddleware) {
+      const sourceContentsStream = streamingSourceContentsCache.stream(
+        replayClient,
+        preferredFrame.location.sourceId
+      );
+
+      await sourceContentsStream.resolver;
+
+      if (sourceContentsStream.value) {
+        if (isReduxMiddleware(sourceContentsStream.value, preferredFrame.location)) {
+          isMiddleware = true;
+          currentPreferredIndex++;
+          preferredFrame = filteredPauseFrames[currentPreferredIndex];
+        } else {
+          isMiddleware = false;
+        }
+      }
+    }
+
+    const frameSteps = await frameStepsCache.readAsync(
+      replayClient,
+      pauseId,
+      preferredFrame.protocolId
+    );
+
+    const pointsWithLocations =
+      frameSteps?.flatMap(step => {
+        return step.frame
+          ?.map(l => {
+            return {
+              location: l,
+              point: step,
+            };
+          })
+          .filter(Boolean) as PointWithLocation[];
+      }) ?? [];
+
+    const preferredLocation = getPreferredLocation(sourcesState, [preferredFrame.location]);
+
+    // One of these locations should match up
+    const matchingFrameStep: PointWithLocation | undefined = pointsWithLocations.find(step => {
+      // Intentionally ignore columns for now - this seems to produce better results
+      // that line up with the hit points in a print statement
+      return (
+        step.location.sourceId === preferredLocation.sourceId &&
+        step.location.line === preferredLocation.line
+      );
+    });
```

Comment:
Now that this function is getting more complex, it might be nice to pull this logic for getting the matching  frame step into a util


```js
function getMatchingFrameStep(frameSteps, preferredLocation) {
  ...
}
```

btw i'm not sure if this is the best API.

Reply:
FWIW the _real_ answer here is to implement https://linear.app/replay/issue/BAC-2915/stack-frames-should-come-with-execution-points and stop trying to calculate this ourselves.

Reply:
I've moved it into a separate util for now since `ReactPanel.tsx` and `ReduxDevtools.tsx` are doing the same thing but yes having frames have their execution point would make it _alot_ simpler